Add gameboard view and run_test

apojomovsky opened the pull request:
Fixes #23

apojomovsky committed:
Created simple view

apojomovsky committed:
Removed show from Gameboard; Created gameboard_view class

apojomovsky committed:
Created run_test.py file

basicNew commented on game/gameboard_view.py:
```diff
@@ -0,0 +1,13 @@
+
+class GameBoardView(object):
+
+    def show(self, board):
```
I think it would be better to have the board as part of the constructor and then the `show` take no params.

basicNew commented on game/gameboard_view.py:
```diff
@@ -0,0 +1,13 @@
+    def show(self, board):
+        """Print the board of a given gameboard in a nice format"""
+        print(chr(27) + "[2J")
+        for row in board.retrieve_matrix():
```
`board.retrieve_matrix()` doesn't look good. I thought that the idea was for the board to encapsulate its internal implementation. We discussed this at some point that you would need to provide, either access per row / column or some sort of iterator. Let's discuss this tomorrow.

apojomovsky replied:
Added a new method: get_matrix_after_doing, where I return a matrix (list of lists), which is an image of the original matrix after applying a given function

basicNew reviewed (commented):
Left a couple of comments

apojomovsky committed:
Added method that iterates over the matrix and applies a function to each item

apojomovsky committed:
Added get_matrix_after_doing method to gameboard

apojomovsky commented:
Just addressed your comments @basicNew, this is ready for another pass

apojomovsky committed:
Added rate argument to run_game

basicNew commented on game/gameboard.py:
```diff
@@ -116,6 +116,16 @@ def is_game_over(self):
     def retrieve_matrix(self):
         return self._matrix
 
+    def get_matrix_after_doing(self, function = lambda x: x):
```
Sorry, but  I found this really confusing. You are using the same method to iterate / map the board and generating a copy of it. Plus the name when it has no argument is really weird (`board.get_matrix_after_doing()` - after doing what?). I would suggest:

- If you want to iterate the board, make it an iterable object (see e.g. http://stackoverflow.com/questions/19151/build-a-basic-python-iterator).
- If you want to map it to another thing (and it is not already an iterator) define a map function that forces the caller to send the lambda. 
- If you want the matrix, define a `get_matrix()`. But as we discussed before, I really think that defining a way to iterate on the object is a better approach.

apojomovsky committed:
Added iterators for accessing to members of the gameboard matrix

apojomovsky commented:
Added the public method "get_rows" (and a unit test for it) that returns an iterator, also added _get_columns, _get_diagonals and _get_entries, which are only being used as part of the internal implementation for now, but I'll make them public in the future (and also add the respective tests for them), when adding strategy four and five (WIP)

basicNew commented on game/gameboard.py:
```diff
@@ -239,23 +265,25 @@ def _winner_in_array(self, testing_array):
-            if self._winner_in_array(self._matrix.diagonal(i)) or \
-               self._winner_in_array(numpy.flipud(self._matrix).diagonal(i)):
+        diag_iter = self._get_diagonals()
+        for diagonal in diag_iter:
```
I really don't see the need for the `*_iter` variables i the methods above... Why not just `for diagonal in self._get_diagonals():`?

basicNew reviewed (commented):
Just a minor comment. Feel free to merge after fixing that

apojomovsky committed:
Removed unneeded iterator variables
